### ISSUE TITLE
Tss2_Sys_SetCmdAuths issue

### ISSUE THREAD

**Issue by juliantrzeciak** [01org/TPM2.0-TSS#51]
file: https://github.com/01org/TPM2.0-TSS/blob/master/sysapi/sysapi/authorizations.c 
commit: https://github.com/01org/TPM2.0-TSS/commit/ca8e7f3c705553f95e6ecafe96b291ab57dbb71f)

**1. TSS2_SYS_RC_BAD_VALUE should cause immediate return**
this is "source null pointer" scenario
line 78 change **break** to **return rval**

**Comment by wcarthur1** [comment]
Julian,

I've added tests and fixed all the issues with SetCmdAuths. 
Can you look at my pull request #65 and approve them for merging?

Thanks,
Will

**Comment by juliantrzeciak** [comment]
Will,

I'm affaid this one is not fixed.

Scenario is that in 
`
 TSS2_RC Tss2_Sys_SetCmdAuths(
     TSS2_SYS_CONTEXT            *sysContext,
     const TSS2_SYS_CMD_AUTHS 	*cmdAuthsArray
     )
`
 
the case for cmdAuthsArray==NULL is covered in line num.40

But I was thinking of kind more weird scenario when null-pointer for one of auths member within the TSS2_SYS_CMD_AUTHS 

`
// Input structure for authorization area(s).
typedef struct {
    uint8_t cmdAuthsCount;
    TPMS_AUTH_COMMAND **cmdAuths;
} TSS2_SYS_CMD_AUTHS;
`

I'm sorry for confusion.

now have a some pointer to TSS2_SYS_CMD_AUTHS structure but first of TPMS_AUTH_COMMAND members is null then on line 78 break out of 72-92 'for loop' then set auth from a null member, then disaster.

Best Regards,
Julian

**Comment by wcarthur1** [comment]
Julian,

I looked at this, and initially I agreed with you.

But then I was trying to figure out why my test case didn’t catch this.

What will happen when I break out of the loop is that I will calculate the size of any valid auth structures found before the NULL one (in the case you mention of the first one being NULL, there would be no valid auth structures.  Then I will move existing parameters out of the way to make room for the auth structures (line 106).  Then at line 108, I will see the BAD_VALUE rval and skip the rest and exit.

I do agree that it would be clearer if I checked the rval after exiting the for loop and handled it there.  As soon as I get a chance, I will make that change.

Will Arthur
Intel Corporation
Server Security Firmware
803-216-2101
